### PR TITLE
Adds in to IF

### PR DESCRIPTION
**Pull request by RaphaelAlvez**
Just for consistency with Select both have the same relations now

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3454/pr-test (or any other room on that server)

File: client/js/jsonedit.js [hunk 1]
```diff
@@ -464,12 +464,11 @@ function jeAddCommands() {
   jeAddEnumCommands('^.*\\(CLICK\\) ↦ mode', [ 'respect', 'ignoreClickable', 'ignoreClickRoutine', 'ignoreAll' ]);
   jeAddEnumCommands('^.*\\(FLIP\\) ↦ faceCycle', [ 'forward', 'backward', 'random' ]);
   jeAddEnumCommands('^.*\\(GET\\) ↦ aggregation', [ 'first', 'last', 'array', 'average', 'median', 'min', 'max', 'sum' ]);
-  jeAddEnumCommands('^.*\\(IF\\) ↦ relation', [ '<', '<=', '==', '!=', '>', '>=' ]);
   jeAddEnumCommands('^.*\\(LABEL\\) ↦ mode', [ 'set', 'dec', 'inc', 'append' ]);
   jeAddEnumCommands('^.*\\(ROTATE\\) ↦ angle', [ 45, 60, 90, 135, 180 ]);
   jeAddEnumCommands('^.*\\(ROTATE\\) ↦ mode', [ 'set', 'add' ]);
   jeAddEnumCommands('^.*\\(SELECT\\) ↦ mode', [ 'set', 'add' ]);
-  jeAddEnumCommands('^.*\\(SELECT\\) ↦ relation', [ '<', '<=', '==', '!=', '>', '>=', 'in' ]);
+  jeAddEnumCommands('^.*(\\(SELECT\\)||\\(IF\\)) ↦ relation', [ '<', '<=', '==', '!=', '>', '>=', 'in' ]);
```
Comment (larkob) [inline client/js/jsonedit.js:471]: shouldn't this be a single `|` - as this is a regular expression?

Comment (RaphaelAlvez) [reply client/js/jsonedit.js]: fixed

**Comment by bjalder26** [comment]
I set up a button and some labels, in the test room above, so that the button tests IF an input is IN a set.  Seems to work great for numbers, but not work for letters.  Maybe I am doing something wrong though.

**Comment by RaphaelAlvez** [comment]
> I set up a button and some labels, in the test room above, so that the button tests IF an input is IN a set. Seems to work great for numbers, but not work for letters. Maybe I am doing something wrong though.

thanks for the feedback. I did in fact tested it with number. I'll take look

**Comment by bjalder26** [comment]
> > I set up a button and some labels, in the test room above, so that the button tests IF an input is IN a set. Seems to work great for numbers, but not work for letters. Maybe I am doing something wrong though.
> 
> thanks for the feedback. I did in fact tested it with number. I'll take look

You're welcome, thanks for working on this.  I was typing up how I tested it and it worked, then I thought... what if it was letters.  I found some other strange occurrences as well.  Larger sets of numbers seem to wind up true for numbers not included, and a mix of letters and numbers winds up the same way.  I left examples in the room.

**Comment by bjalder26** [comment]
I think I figured out what it is doing.  It appears to be checking if the index exists rather than the value of that index.

**Comment by rogerl50** [comment]
I think that `in` doesn't work the way you thought it did: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in

**Comment by RaphaelAlvez** [comment]
@bjalder26 and @rogerl50 it should be working now

**Comment by RaphaelAlvez** [comment]
> I think that `in` doesn't work the way you thought it did: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in

yeah I got it wrong

**Comment by bjalder26** [comment]
It looks like if/in works to check if a string is contained in another string.  I updated my button in the upper left to create strings rather than an array, and it appears to work now.

**Comment by RaphaelAlvez** [comment]
> It looks like if/in works to check if a string is contained in another string.  I updated my button in the upper left to create strings rather than an array, and it appears to work now.

Yeah that was kind intended.

If you tested, everything works and you agree with the changes please review. If you don't want to check the code you can add a review like "tested and approved"

We need more reviewers and testers. That has been out bottle neck

**Comment by bjalder26** [comment]
> Yeah that was kind intended.
> 
> If you tested, everything works and you agree with the changes please add a review to the PR. If you don't want to check the code you can add a review like "tested and approved"
> 
> We need more reviewers and testers. That has been out bottle neck

I don't understand the Javascript.  However, from my tests, it appears to work exactly as intended.  It seems like an improvement, so I would approve.

**Review by ArnoldSmith86 — APPROVED** [review]
I was hesitant about adding `in` inside `compute` because it already has `indexOf` but I think that'll be useful for expressions as well.

File: client/js/widgets/widget.js [hunk 2]
```diff
@@ -535,6 +535,9 @@ export class Widget extends StateManaged {
           case 'unshift':
             v[o](x);
             break;
+          case 'in':
+            v = y.indexOf(x) != -1;
+            break;
```
Comment (larkob) [inline client/js/widgets/widget.js:540]: User may want to use the actual `in` function from JavaScript which allows to identify whether a certain property exists in an object, so we should check whether this is an array or not.

```suggestion
          case 'in':
            v = (Array.isArray(y)) ? y.indexOf(x) != -1 : x in y
            break;
```

File: client/js/widgets/widget.js [hunk 3]
```diff
@@ -535,6 +535,9 @@ export class Widget extends StateManaged {
           case 'unshift':
             v[o](x);
             break;
+          case 'in':
+            v = (Array.isArray(y)|(typeof y=="string")) ? y.indexOf(x) != -1 : x in y;
```
Comment (larkob) [inline client/js/widgets/widget.js:539]:
```suggestion
            v = (Array.isArray(y) || typeof y == 'string') ? y.indexOf(x) != -1 : x in y;
```
OR needs two ||

File: client/js/widgets/widget.js [hunk 4]
```diff
@@ -797,8 +800,8 @@ export class Widget extends StateManaged {
               return w.get(a.property) >= a.value;
             else if(a.relation === '>')
               return w.get(a.property) > a.value;
-            else if(a.relation === 'in' && Array.isArray(a.value))
-              return a.value.indexOf(w.get(a.property)) != -1;
+            else if(a.relation === 'in' && ((typeof a.value=="string")||(typeof a.value=="object")))
+              return (Array.isArray(a.value)|(typeof a.value=="string")) ? a.value.indexOf(w.get(a.property)) != -1 : w.get(a.property) in a.value
```
Comment (larkob) [inline client/js/widgets/widget.js:804]:
```suggestion
              return (Array.isArray(a.value) || typeof a.value== 'string') ? a.value.indexOf(w.get(a.property)) != -1 : w.get(a.property) in a.value
```
OR needs two ||

**Comment by RaphaelAlvez** [comment]
@larkob maybe I should leave it to you to add this in you new PR as you are already changing everything in that part of the code

**Comment by bjalder26** [comment]
I tested again, and can't find any unexpected behavior.  I even tried looking for a numerical value (not a string) in a string and it worked.  Looking for strings in arrays works, strings in strings, and numbers in arrays or strings.  What you will not find is the string "1" in array such as [1, 2, 3], but that seems like correct behavior to me.

**Comment by RaphaelAlvez** [comment]
this should be included in PR #469

**Pull request closed without merging**